### PR TITLE
GH-34382: [C++] Support more types in run_end_encode and run_end_decode functions

### LINKED ISSUE #34382: [C++] Run-end encoding and decoding: Add support for more types
### Describe the enhancement requested

#34195 introduces two functions `run_end_encode` and `run_end_decode` supporting with support for numeric, null, and bool types. More types can and should be supported.

TODO list:

 - [x] All `NumericTypes()`
 - [x] `bool()`
 - [x] `null()`
 - [x] Temporal, Interval, and Duration types
 - [x] Variable length types
 - [x] Fixed size binary types

### Component(s)

C++

### PR DESCRIPTION
**Pull request by felipecrv**
* Closes: #34382

### CODE REVIEW

**Comment by felipecrv** [comment]
@zeroshade

**Comment by dinimar** [comment]
> @dinimar you asked what was needed to add more types in the issue and after almost 600 lines... here it is :)

@felipecrv Luckily, I didn't start working on this issue. Didn't think it will take so much knowledge about the project)

**Comment by felipecrv** [comment]
CI failure seems unrelated.

File: cpp/src/arrow/compute/kernels/vector_run_end_encode_test.cc [hunk 1]
```diff
@@ -317,7 +320,52 @@ INSTANTIATE_TEST_SUITE_P(
...
+                R"(["abc", "abc", "abc", "def", "def", "def", "ghi", "ghi"])",
+                R"(["abc", "def", "ghi"])", "[3, 6, 8]"),
+            REETestData::JSON(
+                fixed_size_binary(3),
+                R"([null, "abc", "abc", "abc", "def", "def", "def", "ghi", "ghi", null, null])",
+                R"([null, "abc", "def", "ghi", null])", "[1, 4, 7, 9, 11]"),
+            // String and binary types
+            REETestData::JSON(
+                utf8(), R"(["abc", "abc", "", "", "de", "de", "de", "ghijkl", "ghijkl"])",
+                R"(["abc", "", "de", "ghijkl"])", "[2, 4, 7, 9]"),
```
Comment (wjones127) [inline cpp/src/arrow/compute/kernels/vector_run_end_encode_test.cc:356]: Perhaps we can reuse these test cases in a for loop, and also cover the `large_*` type variants?

**Review by wjones127 — CHANGES_REQUESTED** [review]
This looks like a great addition! Have one small suggestion on the tests.

File: cpp/src/arrow/compute/kernels/vector_run_end_encode_test.cc (hunk 1 above)
Comment (felipecrv) [reply cpp/src/arrow/compute/kernels/vector_run_end_encode_test.cc:356]: Is copy and paste OK? Keeps the data neatly declarative here.

Comment (wjones127) [reply cpp/src/arrow/compute/kernels/vector_run_end_encode_test.cc:356]: I was thinking something like:

```cpp
        std::shared_ptr<DataType> types[] = {binary(), large_binary(), utf8(), large_utf8()};
        for (auto binary_type : types) {
            REETestData::JSON(
                binary_type,
                R"(["abc", "abc", "", "", "de", "de", "de", "ghijkl", "ghijkl"])",
                R"(["abc", "", "de", "ghijkl"])", "[2, 4, 7, 9]"),
            REETestData::JSON(
                binary_type,
                R"(["abc", "abc", "", "", "de", "de", "de", null, null, "ghijkl", "ghijkl"])",
                R"(["abc", "", "de", null, "ghijkl"])", "[2, 4, 7, 9, 11]"),
        }
```

So still pretty declarative.

Comment (wjones127) [reply cpp/src/arrow/compute/kernels/vector_run_end_encode_test.cc:356]: Or put another way: I was just speaking of the string / binary cases here.

Comment (felipecrv) [reply cpp/src/arrow/compute/kernels/vector_run_end_encode_test.cc]: Done. Done in the latest two commits.

**Review by wjones127 — APPROVED** [review]

**Pull request merged**